src/lib.rs

```diff
@@ -3,30 +3,45 @@ use kvmi_sys::{kvmi_qemu2introspector, kvmi_introspector2qemu};
 use std::ffi::{CString};
 use std::ptr::{null_mut};
 use std::os::raw::{c_void, c_uchar, c_int};
+use std::sync::{Mutex, Condvar};
 
 
 #[derive(Debug)]
 pub struct KVMi {
     ctx: *mut c_void,
+    dom: *mut c_void,
+}
+
+#[derive(Debug)]
+struct KVMiCon {
+    dom: *mut c_void,
+    guard: Mutex<bool>,
+    condvar: Condvar,
 }
 
 extern "C" fn new_guest_cb(dom: *mut c_void,
                            uuid: *mut [c_uchar; 16usize],
-                           ctx: *mut c_void) -> c_int {
+                           cb_ctx: *mut c_void) -> c_int {
     println!("new guest cb !");
+    let kvmi_con = unsafe { &mut *(cb_ctx as *mut KVMiCon) };
+    let mut started = kvmi_con.guard.lock().unwrap();
```

Comment:
Lots of `unwrap` :(
You must ensure that the same thread cannot call acquire lock twice, or this will fail.
For debugging purpose, you may also want to use `expect("a good error msg")` instead of unwrap